dask/array/_shuffle.py

```diff
@@ -0,0 +1,199 @@
+    if not isinstance(indexer, list) or not all(isinstance(i, list) for i in indexer):
+        raise ValueError("indexer must be a list of lists of positional indices")
+
+    if not axis <= len(chunks):
+        raise ValueError(
+            f"Axis {axis} is out of bounds for array with {len(chunks)} axes"
+        )
+
+    if max(map(max, indexer)) >= sum(chunks[axis]):
+        raise IndexError(
+            f"Indexer contains out of bounds index. Dimension only has {sum(chunks[axis])} elements."
+        )
+
+    chunksize_tolerance = config.get("array.shuffle.chunksize-tolerance")
+    average_chunk_size = int(
+        sum(chunks[axis]) / len(chunks[axis]) * chunksize_tolerance
+    )
+
+    # Figure out how many groups we can put into one chunk
+    current_chunk, new_chunks = [], []
+    for idx in indexer:
+        if (
+            len(current_chunk) + len(idx) > average_chunk_size
+            and len(current_chunk) > 0
+        ):
+            new_chunks.append(current_chunk)
+            current_chunk = idx.copy()
+        else:
+            current_chunk.extend(idx)
+            if len(current_chunk) > average_chunk_size / 1.25:
+                new_chunks.append(current_chunk)
+                current_chunk = []
+    if len(current_chunk) > 0:
+        new_chunks.append(current_chunk)
+
+    chunk_boundaries = np.cumsum(chunks[axis])
+
+    # Get existing chunk tuple locations
+    chunk_tuples = list(
+        product(*(range(len(c)) for i, c in enumerate(chunks) if i != axis))
+    )
+
+    intermediates = dict()
+    merges = dict()
+    split_name = f"shuffle-split-{token}"
+    slices = [slice(None)] * len(chunks)
+    split_name_suffixes = count()
+
+    old_blocks = np.empty([len(c) for c in chunks], dtype="O")
+    for old_index in np.ndindex(old_blocks.shape):
+        old_blocks[old_index] = (in_name,) + old_index
+
+    for new_chunk_idx, new_chunk_taker in enumerate(new_chunks):
+        new_chunk_taker = np.array(new_chunk_taker)
+        sorter = np.argsort(new_chunk_taker)
+        sorted_array = new_chunk_taker[sorter]
+        source_chunk_nr, taker_boundary = np.unique(
+            np.searchsorted(chunk_boundaries, sorted_array, side="right"),
+            return_index=True,
+        )
+        taker_boundary = taker_boundary.tolist()
+        taker_boundary.append(len(new_chunk_taker))
+
+        taker_cache = {}
+        for chunk_tuple in chunk_tuples:
+            merge_keys = []
+
+            for c, b_start, b_end in zip(
+                source_chunk_nr, taker_boundary[:-1], taker_boundary[1:]
+            ):
+                # insert our axis chunk id into the chunk_tuple
+                chunk_key = convert_key(chunk_tuple, c, axis)
+                name = (split_name, next(split_name_suffixes))
+                this_slice = slices.copy()
+
+                # Cache the takers to allow de-duplication when serializing
+                # Ugly!
```

Comment:
```suggestion
```

Reply:
Hm I wanted to express that this is necessary but very ugly, so this is kind of intended

Reply:
Ah, makes sense.